apache/cxf
# CXF-8817: Micrometer Observation

This PR adds support for Micrometer Observation instrumentation of Apache CXF.

NOTE: This PR is heavily influenced by Brave's implementation.

Dependencies
- Adds micrometer tracing
- Adds zipkin sender

CXF metrics project
- Now also does tracing (TODO: maybe this should be done in a separate module?)
- Adds dependency on `rt-management` since metrics will now also be related to tracing 
- Adds dependency to Micrometer Observation
- Is a copy of the Brave Tracing module but adopted to Micrometer Observation API
- Uses low and high cardinality tags basing on OpenTelemetry Semantic Conventions for RPC and HTTP (JaxRs) from 19.07.2023. VERY IMPORTANT: This doesn't mean that this project is OTel Semantic Convention compliant!!!
- Uses `ObservationDocumentation` that using https://micrometer.io/docs/observation#_documentation_generation could generate asciidoctor documentation of observability data

Spring Boot
- adds `MicrometerObservationAutoConfiguration` that registers `Feature`s as beans

Tests
- reused existing Brave tests with Micrometer Observation, altered existing tagging and naming defaults to make the test pass with Brave tagging and naming defaults
- reused existing Boot tests to additionally assert observation based metrics and traces

TODO ? :
- maybe the code shouldn't be in CXF metrics module but a completely separate one?
- add Javadocs?
- add documentation - where should it be added?

Screenshots from Zipkin:

![Screenshot from 2023-07-19 13-25-48](https://github.com/apache/cxf/assets/3297437/54779228-91a2-4c53-8339-60f13d684d03)

![Screenshot from 2023-07-19 13-25-55](https://github.com/apache/cxf/assets/3297437/3b25c834-f45a-418d-aead-4ef6428583e8)

Fixes https://issues.apache.org/jira/browse/CXF-8817
